### PR TITLE
Fully implement DistributionMeasure and add var transforms

### PR DESCRIPTION
**Pull request by oschulz**

### CODE REVIEW

**Comment by oschulz** [comment]
Passes tests locally, will need a new release of MeasureBase to run CI.

File: src/measure_interface.jl [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+@inline MeasureBase.unsafe_logdensityof(d::Distribution, x) = DensityInterface.logdensityof(d, x)
+
+@inline MeasureBase.insupport(d::Distribution, x) = Distributions.insupport(d, x)
+
+@inline MeasureBase.basemeasure(d::Distribution{<:ArrayLikeVariate{0},<:Continuous}) = Lebesgue()
+@inline MeasureBase.basemeasure(d::Distribution{<:ArrayLikeVariate,<:Continuous}) = Lebesgue()^size(d)
+@inline MeasureBase.basemeasure(d::Distribution{<:ArrayLikeVariate{0},<:Discrete}) = Counting()
+@inline MeasureBase.basemeasure(d::Distribution{<:ArrayLikeVariate,<:Discrete}) = Counting()^size(d)
+
+@inline MeasureBase.insupport(d::Distribution) = Distributions.insuinsupportpport(d)
```
Comment (cscherrer) [inline src/measure_interface.jl]:
```suggestion
@inline MeasureBase.insupport(d::Distribution, x) = Distributions.insupport(d, x)
```

File: src/utils.jl [hunk 2]
```diff
@@ -1,5 +1,32 @@
...
+Convert x to use `T` as it's underlying type for real numbers.
+"""
+function convert_realtype end
+
+_convert_realtype_pullback(ΔΩ) = NoTangent(), NoTangent, ΔΩ
+ChainRulesCore.rrule(::typeof(convert_realtype), ::Type{T}, x) where T = convert_realtype(T, x), _convert_realtype_pullback
+
+@inline convert_realtype(::Type{T}, x::T) where {T<:Real} = x
+@inline convert_realtype(::Type{T}, x::AbstractArray{T}) where {T<:Real} = x
+@inline convert_realtype(::Type{T}, x::U) where {T<:Real,U<:Real} = T(X)
```
Comment (cscherrer) [inline src/utils.jl]:
```suggestion
@inline convert_realtype(::Type{T}, x::U) where {T<:Real,U<:Real} = T(x)
```

File: src/univariate.jl [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    @eval begin
+        @inline MeasureBase.transport_origin(::$A) = $B()
+        @inline MeasureBase.to_origin(ν::$A, y) = _affine_to_origin(ν, y)
+        @inline MeasureBase.from_origin(ν::$A, x) = _origin_to_affine(ν, x)
+    end
+end
+
+@inline MeasureBase.transport_origin(::Exponential) = StdExponential()
+@inline MeasureBase.to_origin(ν::Exponential, y) = scale(ν) \ y
+@inline MeasureBase.from_origin(ν::Exponential, x) = scale(ν) * x
```
Comment (cscherrer) [inline src/univariate.jl]: `scale` is not defined

File: src/measure_interface.jl [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+@inline MeasureBase.insupport(d::Distribution, x) = Distributions.insupport(d, x)
+
+@inline MeasureBase.basemeasure(d::Distribution{<:ArrayLikeVariate{0},<:Continuous}) = Lebesgue()
+@inline MeasureBase.basemeasure(d::Distribution{<:ArrayLikeVariate,<:Continuous}) = Lebesgue()^size(d)
+@inline MeasureBase.basemeasure(d::Distribution{<:ArrayLikeVariate{0},<:Discrete}) = Counting()
+@inline MeasureBase.basemeasure(d::Distribution{<:ArrayLikeVariate,<:Discrete}) = Counting()^size(d)
+
+@inline MeasureBase.insupport(d::Distribution, x) = Distributions.insupport(d, x)
+@inline MeasureBase.paramnames(d::Distribution) = propertynames(d)
+@inline MeasureBase.params(d::Distribution) = NamedTuple{MeasureBase.paramnames(m.d)}(Distributions.params(d))
```
Comment (cscherrer) [inline src/measure_interface.jl]: What's `m` supposed to be here?

File: src/univariate.jl (hunk 3 above)
Comment (oschulz) [reply src/univariate.jl]: It is, this is `Distributions.Exponential`:

```julia
julia> d = Distributions.Exponential(4.2)
Exponential{Float64}(θ=4.2)

julia> scale(d)
4.2
```

Comment (cscherrer) [reply src/univariate.jl]: I think it needs to be brought into scope:
```julia
help?> DistributionMeasures.scale
  No documentation found.

  Binding DistributionMeasures.scale does not exist.
```

Comment (cscherrer) [reply src/univariate.jl]: So we need to either qualify it as `Distributions.scale` or `using Distributions: scale`

File: src/measure_interface.jl (hunk 4 above)
Comment (oschulz) [reply src/measure_interface.jl]: Ah, that's wrong - fixed and added tests.

File: src/standard_dist.jl [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+Base.convert(::Type{D}, d::StandardDist{D,0}) where {D<:Distribution{Univariate,Continuous}} = D(d)
+Base.convert(::Type{Distributions.Product}, d::StandardDist{D,1}) where {D} = Distributions.Product(d)
+
+
+function _checkvarsize(d::Distribution{ArrayLikeVariate{N}}, x::AbstractArray{<:Real,N}) where N
+    size(d) == size(x) || throw(DimensionMismatch("Size of variate doesn't match distribution"))
+end
+
+function ChainRulesCore.rrule(::typeof(_checkvarsize), d::Distribution{ArrayLikeVariate{N}}, x::AbstractArray{<:Real,N}) where N
+    _checkvarsize(d, x), _nogradient_pullback2
```
Comment (cscherrer) [inline src/standard_dist.jl]: Where is `_nogradient_pullback2`?

File: src/univariate.jl [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+
+# Transform between univariate and single-element standard multivariate
+
+function MeasureBase.transport_def(ν::Distribution{Univariate}, μ::StandardDist{D,1}, x) where D
+    return transport_def(ν, StandardDist{D,0}(), only(x))
+end
+
+function MeasureBase.transport_def(ν::StandardDist{D,1}, μ::Distribution{Univariate}, x) where D
+    return Fill(transport_def(StandardDist{D,0}(), μ, only(x)), map(length, ν.axes)...)
```
Comment (cscherrer) [inline src/univariate.jl]: Is `ν.axes` defined?

File: src/univariate.jl (hunk 3 above)
Comment (oschulz) [reply src/univariate.jl]: Done

File: src/univariate.jl (hunk 6 above)
Comment (oschulz) [reply src/univariate.jl]: Ooops, thanks -fixed.

File: src/standard_dist.jl (hunk 5 above)
Comment (oschulz) [reply src/standard_dist.jl]: Ah, thanks for the catch, that was a leftover. I've removed _checkvarsize now in favor of our new `checked_var`.

**Review by cscherrer — APPROVED** [review]

**Comment by oschulz** [comment]
We should wait for cscherrer/MeasureBase.jl#63 before merging and tagging.

**Pull request merged**